src/Microsoft.Health.Dicom.Core/Features/Store/StoreDatasetValidator.cs

```diff
@@ -180,10 +185,12 @@ string EnsureRequiredTagIsPresent(DicomTag dicomTag)
         }
     }
 
-    private void ValidateAllItems(
+    private async Task ValidateAllItems(
         DicomDataset dicomDataset,
         StoreValidationResultBuilder validationResultBuilder)
     {
+        IReadOnlyCollection<QueryTag> queryTags = await _queryTagService.GetQueryTagsAsync();
+        var indexableTags = queryTags.Select(t => t.Tag);
```

Comment:
You should materialize the select and convert it into hashset.

For loop below will keep materializing it each loop on contains